tendermint/src/lite/verifier.rs

```diff
@@ -184,98 +119,118 @@ where
     C: Commit,
     L: TrustThreshold,
-    S: Store<TrustedState = TS>,
-    R: Requester<SignedHeader = SH, ValidatorSet = C::ValidatorSet>,
 {
-    // can we trust the still untrusted header based on the given trusted state?
-    match check_support(
-        trusted_state,
+    // validate the untrusted header against its commit, vals, and next_vals
+    let untrusted_header = untrusted_sh.header();
+    let untrusted_commit = untrusted_sh.commit();
+    validate_untrusted(
         untrusted_header,
-        trust_threshold,
-        trusting_period,
-        now,
-    ) {
-        Ok(_) => {
-            let untrusted_next_vals =
-                req.validator_set(untrusted_header.header().height().increment())?;
-            let untrusted_state = TS::new(untrusted_header, &untrusted_next_vals);
-            store.add(&untrusted_state)?;
-            return Ok(());
-        }
-        Err(e) => {
-            if e != Error::InsufficientVotingPower {
-                return Err(e);
+        untrusted_commit,
+        untrusted_vals,
+        untrusted_next_vals,
+    )?;
+
+    // ensure the new height is higher.
+    // if its +1, ensure the vals are correct.
+    // if its >+1, ensure we can skip to it
+    let trusted_header = trusted_state.last_header().header();
+    let trusted_height = trusted_header.height();
+    let untrusted_height = untrusted_sh.header().height();
+    match untrusted_height.cmp(&trusted_height.increment()) {
+        Ordering::Less => return Err(Error::NonIncreasingHeight),
+        Ordering::Equal => {
+            let trusted_vals_hash = trusted_header.next_validators_hash();
+            let untrusted_vals_hash = untrusted_header.validators_hash();
+            if trusted_vals_hash != untrusted_vals_hash {
+                // TODO: more specific error
+                // ie. differentiate from when next_vals.hash() doesnt
+                // match the header hash ...
+                return Err(Error::InvalidNextValidatorSet);
             }
         }
+        Ordering::Greater => {
+            let trusted_vals = trusted_state.validators();
+            verify_commit_trusting(trusted_vals, untrusted_commit, trust_threshold)?;
+        }
     }
 
-    // Here we can't trust the passed in untrusted header based on the known trusted state.
-    // Run bisection: try again with a pivot header whose height is in the
-    // middle of the trusted height and the desired height (h2.height).
-
-    let trusted_header = trusted_state.last_header();
-    let trusted_height: u64 = trusted_header.header().height().value();
-    let untrusted_height: u64 = untrusted_header.header().height().value();
-    let pivot: u64 = (trusted_height + untrusted_height) / 2;
-    let pivot_header = req.signed_header(pivot)?;
-    let pivot_vals = req.validator_set(pivot)?;
+    // All validation passed successfully. Verify the validators correctly committed the block.
+    verify_commit_full(untrusted_vals, untrusted_sh.commit())
```

Comment:
I would move this after `validate_untrusted()` and even combine them as they perform signed header validation. 
Also isn't `verify_commit_full()` same as `verify_commit_trusting()` with different `trust_threshold`.

Reply:
Yeah, actually something is not right. First, I think it is important that `verify_single()` does both `validate_untrusted()` and `verify_commit_full()` first. Any errors from these functions are fatal and bisection should not proceed. As it is right now, `verify_single()` may return early with error from `verify_commit_trusting()` which causes the bisection caller to start bisecting even if this header/commit is not fully verified. For the second issue, see next comment...

Reply:
So I was actually thinking the opposite, which was that we shouldn't bother trying to fully verify a future header (which is expensive) until we know we can skip to it.

I suppose it's a question of compute vs bandwidth. But even if we did the verification first, a full node who wanted to trick us into using a lot of bandwidth could always give us headers that are well signed by themselves so we'd waste time verifying them and not find out until much later that they're not actually valid from previous ones. So it doesn't really make sense to me to be verifying the commit of a header until we know we can skip to it and that was one of the key points I wanted to effect in these changes. 

Sorry, I should have written that up more explicitly somewhere.

Reply:
And yes, verify_commit_full and _trusting are the same now except the later uses the threshold trait and the former is just fixed to 2/3

Reply:
> So I was actually thinking the opposite, which was that we shouldn't bother trying to fully verify a future header (which is expensive) until we know we can skip to it.

yes, but the problem is that if we know we can skip it, we start recursive bisection  which involves downloading and fully verifying all the pivots...so both bandwidth and cpu.

Reply:
as discussed, this is ok, because the in between headers are being verified fully and hence trusted, even if the node turns out to be bad at the end of the run, so that bandwidth and cpu isn't really wasted.